File: problems/Subsets/subsets.cs
public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        var result = new List<IList<int>>();
        var n = nums.Length;

        for (var k = 0; n >= k; ++k) {
            backtrack(new List<int>(), nums, result, k);
        }

        return result;
    }

    private void backtrack(List<int> curr, int[] nums, IList<IList<int>> res, int count, int idx = 0) {
        // if the combination is done
        if (curr.Count == count) {
            res.Add(curr.ToList());
        }

        for (var i = idx; nums.Length > i; ++i) {
            // add i into the current combination
            curr.Add(nums[i]);
            // use next integers to complete the combination
            backtrack(curr, nums, res, count, 1 + i);
            // backtrack
            curr.RemoveAt(curr.Count - 1);
        }
    }
}

File: problems/Subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();

        for (int idx = 0; nums.length >= idx; ++idx) {
            backtrack(new ArrayList<>(), nums, result, idx, 0);
        }

        return result;
    }

    private void backtrack(List<Integer> currSubset, int[] nums, List<List<Integer>> result, int count, int idx) {
        if (count == idx) {
            result.add(currSubset);
        }

        for (int i = idx; nums.length > i; ++i) {
            currSubset.add(nums[i]);
            backtrack(new ArrayList<>(currSubset), nums, result, count, 1 + i);
            currSubset.remove(currSubset.size() - 1);
        }
    }
}
